src/interpreter.rs

```diff
@@ -18,6 +78,68 @@ impl Execute for Expression {
     }
 }
 
+#[derive(Debug)]
+pub enum OperandItem<'a> {
+    Val(bool),
+    Predicate(&'a Predicate),
+}
+fn evaluate_operand_item(item: OperandItem, ctx: &mut Context, m: &mut Match) -> bool {
+    match item {
+        OperandItem::Val(b) => b,
+        OperandItem::Predicate(p) => p.execute(ctx, m),
+    }
+}
+
+impl Execute for Route {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        let mut arr_predicate: ArrayBuf<OperandItem, 2> = ArrayBuf::new();
+
+        for item in &self.stack {
+            match item {
+                RouteTerm::LogicalOperator(op) => {
+
+                    match op {
+                        RouteLogicalOperators::And => {
+                            let left = evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
+                            if left == false {
+                                // short circuit
+                                arr_predicate.pop();
+                                arr_predicate.push(OperandItem::Val(false));
+                            } else {
+                                let right =
+                                    evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
+                                arr_predicate.push(OperandItem::Val(right));
+                            }
+                        }
+                        RouteLogicalOperators::Or => {
+                            let left = evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
+                            if left == true {
+                                // short circuit
+                                arr_predicate.pop();
+                                arr_predicate.push(OperandItem::Val(true));
+                            } else {
+                                let right =
+                                    evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
+                                arr_predicate.push(OperandItem::Val(right));
+                            }
+                        }
+                        RouteLogicalOperators::Not => {
+                            let operand =
+                                evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
+                            arr_predicate.push(OperandItem::Val(!operand));
+                        }
+                    }
+                }
+                RouteTerm::Predicate(p) => {
+
+                    arr_predicate.push(OperandItem::Predicate(p));
+                }
+            }
+        }
+        return evaluate_operand_item(arr_predicate.pop().unwrap(), ctx, m);
```

Comment:
Should we check if this is the last item in the stack?

Reply:
Fixed. thanks.